## Part I. Preliminary Data Analysis

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
train_1 = pd.read_csv('ACT1_train_450.csv', dtype={"MOLECULE": object, "Act": float})

In [3]:
train_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29792 entries, 0 to 29791
Columns: 443 entries, Unnamed: 0 to Act
dtypes: float64(1), int64(442)
memory usage: 100.7 MB


In [4]:
train_1.head(5)

Unnamed: 0  D_212  D_213  D_214  D_215  D_216  D_217  D_219  D_294  D_295  \
0        3630      0      0      0      0      0      0      0      0      0   
1       30103      0      0      0      0      0      0      0      1      2   
2        8337      1      1      2      0      0      0      0      2      1   
3       24497      0      1      1      1      0      0      0      2      1   
4       20668      0      0      0      0      0      0      0      3      2   

    ...    D_7351  D_7352  D_7954  D_7955  D_7957  D_7958  D_8051  D_8053  \
0   ...         0       2       0       0       2       1       1       4   
1   ...         1       0       0       0       0       0       0       0   
2   ...         0       0       0       0       0       0       0       0   
3   ...         0       0       0       0       1       1       0       2   
4   ...         0       0       0       0       0       0       0       0   

   D_8054     Act  
0       1  4.8624  
1       0  4.9817  
2       0  5.2265  
3       0  6.2863  
4       0  4.5891  

[5 rows x 443 columns]

In [5]:
train_1 = train_1.dropna(axis = 1) # drop all columns that contain null value

In [6]:
train_1.head(5)

Unnamed: 0  D_212  D_213  D_214  D_215  D_216  D_217  D_219  D_294  D_295  \
0        3630      0      0      0      0      0      0      0      0      0   
1       30103      0      0      0      0      0      0      0      1      2   
2        8337      1      1      2      0      0      0      0      2      1   
3       24497      0      1      1      1      0      0      0      2      1   
4       20668      0      0      0      0      0      0      0      3      2   

    ...    D_7351  D_7352  D_7954  D_7955  D_7957  D_7958  D_8051  D_8053  \
0   ...         0       2       0       0       2       1       1       4   
1   ...         1       0       0       0       0       0       0       0   
2   ...         0       0       0       0       0       0       0       0   
3   ...         0       0       0       0       1       1       0       2   
4   ...         0       0       0       0       0       0       0       0   

   D_8054     Act  
0       1  4.8624  
1       0  4.9817  
2       0  5.2265  
3       0  6.2863  
4       0  4.5891  

[5 rows x 443 columns]

There is no missing values in train_1.

In [7]:
train_1.describe()

Unnamed: 0         D_212         D_213         D_214         D_215  \
count  29792.000000  29792.000000  29792.000000  29792.000000  29792.000000   
mean   18626.388494      0.435117      0.547697      0.560755      0.569851   
std    10748.385829      1.056851      1.282340      1.289464      1.196240   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%     9350.750000      0.000000      0.000000      0.000000      0.000000   
50%    18607.500000      0.000000      0.000000      0.000000      0.000000   
75%    27941.250000      0.000000      0.000000      1.000000      1.000000   
max    37240.000000     12.000000     14.000000     16.000000     14.000000   

              D_216         D_217         D_219         D_294         D_295  \
count  29792.000000  29792.000000  29792.000000  29792.000000  29792.000000   
mean       0.539507      0.510875      0.369965      1.526316      1.844623   
std        1.162756      1.122913      1.040328      1.645797      2.109708   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      1.000000      1.000000   
75%        1.000000      1.000000      0.000000      2.000000      3.000000   
max       15.000000     14.000000     12.000000     16.000000     18.000000   

           ...             D_7351        D_7352        D_7954        D_7955  \
count      ...       29792.000000  29792.000000  29792.000000  29792.000000   
mean       ...           0.603921      0.429713      0.750134      0.577773   
std        ...           1.096166      1.002109      1.305478      1.111365   
min        ...           0.000000      0.000000      0.000000      0.000000   
25%        ...           0.000000      0.000000      0.000000      0.000000   
50%        ...           0.000000      0.000000      0.000000      0.000000   
75%        ...           1.000000      0.000000      1.000000      1.000000   
max        ...          11.000000     11.000000     11.000000      9.000000   

             D_7957       D_7958        D_8051        D_8053        D_8054  \
count  29792.000000  29792.00000  29792.000000  29792.000000  29792.000000   
mean       0.544811      0.56250      0.581733      0.655277      0.586970   
std        1.086445      1.13118      1.012741      1.089153      1.071895   
min        0.000000      0.00000      0.000000      0.000000      0.000000   
25%        0.000000      0.00000      0.000000      0.000000      0.000000   
50%        0.000000      0.00000      0.000000      0.000000      0.000000   
75%        1.000000      1.00000      1.000000      1.000000      1.000000   
max       11.000000     13.00000      8.000000      8.000000      8.000000   

                Act  
count  29792.000000  
mean       4.691989  
std        0.650919  
min        4.300300  
25%        4.300300  
50%        4.300300  
75%        4.907600  
max        8.134000  

[8 rows x 443 columns]

## Part II. Machine Learning

Split train data into train/test part (80:20)

In [8]:
from sklearn.model_selection import train_test_split
import numpy as np
y = train_1['Act'].values
train_1 = train_1.drop(['Act', 'Unnamed: 0'], axis = 1)
x = train_1.values
X_train, X_test, Y_train, Y_test = train_test_split(x, y, train_size = 0.80)



C:\Users\willi\Anaconda3\envs\data-x\lib\site-packages\sklearn\model_selection\_split.py:2179: FutureWarning: From version 0.21, test_size will always complement train_size unless both are specified.
  FutureWarning)


In [9]:
print(np.shape(Y_train), np.shape(X_train), np.shape(Y_test), np.shape(X_test))
Y_train = np.reshape(Y_train,(23833,1))
Y_test = np.reshape(Y_test,(5959,1))

(23833,) (23833, 441) (5959,) (5959, 441)


There are 29792 entries for training the model and 7448 entries for model testing. This split gurantees there's enough data for both the training and testing steps.

### 2. Define evaluation metric r^2

In [93]:
def r_square(X, Y):
    """ r^2 value defined by the competition host, r^2 = 1 indicates 100% prediction accuracy
    """
    avx = np.mean(X)
    avy = np.mean(Y)
    sum1, sumx, sumy = 0, 0, 0
    for i in range(len(X)):
        sum1 += (X[i] - avx)*(Y[i] - avy)
        sumx += (X[i] - avx)*(X[i] - avx)
        sumy += (Y[i] - avy)*(Y[i] - avy)
    print(len(X), sum1, sumx, sumy)
    return sum1*sum1/(sumx*sumy)

### 3. Train neural network

#### 1) Optimized NN with two hidden layers

In [11]:
from tensorflow.keras.models import Sequential
import keras
model = Sequential()

Using TensorFlow backend.


In [84]:
from keras import backend as K
K.clear_session()

def coeff_determination(y_true, y_pred):
    SS_res =  K.sum(K.square( y_true-y_pred )) 
    SS_tot = K.sum(K.square( y_true - K.mean(y_true) ) ) 
    return ( 1 - SS_res/(SS_tot + K.epsilon()) )

In [85]:
#Keras model
from tensorflow.keras.layers import Dense
from keras.optimizers import adam 

model = keras.models.Sequential()
model.add(keras.layers.core.Dense(units = 50, activation='relu', input_dim = 441))
model.add(keras.layers.core.Dense(25, activation='relu'))
model.add(keras.layers.core.Dense(1, activation = 'relu'))

adam = adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)

model.compile(loss="mean_squared_error", optimizer=adam, metrics = [coeff_determination])
print(model.summary())


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 50)                22100     
_________________________________________________________________
dense_2 (Dense)              (None, 25)                1275      
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 26        
Total params: 23,401
Trainable params: 23,401
Non-trainable params: 0
_________________________________________________________________
None


In [86]:
model.fit(X_train, Y_train, epochs = 500, batch_size= 300)

Epoch 1/500
23833/23833 [==============================] - 0s 15us/step - loss: 7.0242 - coeff_determination: -15.4794
Epoch 2/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.5637 - coeff_determination: -0.3384
Epoch 3/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.4039 - coeff_determination: 0.0328
Epoch 4/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.3471 - coeff_determination: 0.1686
Epoch 5/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.3024 - coeff_determination: 0.2777
Epoch 6/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.2818 - coeff_determination: 0.3285
Epoch 7/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.2588 - coeff_determination: 0.3805
Epoch 8/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.2456 - coeff_determination: 0.4125
Epoch 9/500
23833/23833 [==============================] - 0s 9us/st

23833/23833 [==============================] - 0s 8us/step - loss: 0.1170 - coeff_determination: 0.7188
Epoch 72/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.1189 - coeff_determination: 0.7158
Epoch 73/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.1106 - coeff_determination: 0.7355
Epoch 74/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.1130 - coeff_determination: 0.7292
Epoch 75/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.1116 - coeff_determination: 0.7317
Epoch 76/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.1109 - coeff_determination: 0.7348
Epoch 77/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.1105 - coeff_determination: 0.7355
Epoch 78/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.1096 - coeff_determination: 0.7378
Epoch 79/500
23833/23833 [==============================] - 0s 9us/step - los

23833/23833 [==============================] - 0s 8us/step - loss: 0.0797 - coeff_determination: 0.8089
Epoch 141/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0816 - coeff_determination: 0.8051
Epoch 142/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0810 - coeff_determination: 0.8072
Epoch 143/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0797 - coeff_determination: 0.8100
Epoch 144/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0785 - coeff_determination: 0.8110
Epoch 145/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0811 - coeff_determination: 0.8053
Epoch 146/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0750 - coeff_determination: 0.8198
Epoch 147/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0790 - coeff_determination: 0.8099
Epoch 148/500
23833/23833 [==============================] - 0s 8us/st

23833/23833 [==============================] - 0s 8us/step - loss: 0.0619 - coeff_determination: 0.8518
Epoch 210/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0625 - coeff_determination: 0.8501
Epoch 211/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0625 - coeff_determination: 0.8503
Epoch 212/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0621 - coeff_determination: 0.8502
Epoch 213/500
23833/23833 [==============================] - 0s 10us/step - loss: 0.0601 - coeff_determination: 0.8556
Epoch 214/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0613 - coeff_determination: 0.8529
Epoch 215/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0604 - coeff_determination: 0.8546
Epoch 216/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0634 - coeff_determination: 0.8480
Epoch 217/500
23833/23833 [==============================] - 0s 8us/s

23833/23833 [==============================] - 0s 9us/step - loss: 0.0522 - coeff_determination: 0.8747
Epoch 279/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0516 - coeff_determination: 0.8762
Epoch 280/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0516 - coeff_determination: 0.8765
Epoch 281/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0533 - coeff_determination: 0.8718
Epoch 282/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0521 - coeff_determination: 0.8749
Epoch 283/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0512 - coeff_determination: 0.8773
Epoch 284/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0524 - coeff_determination: 0.8742
Epoch 285/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0522 - coeff_determination: 0.8751
Epoch 286/500
23833/23833 [==============================] - 0s 9us/st

23833/23833 [==============================] - 0s 9us/step - loss: 0.0456 - coeff_determination: 0.8906
Epoch 348/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0497 - coeff_determination: 0.8796
Epoch 349/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0463 - coeff_determination: 0.8895
Epoch 350/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0466 - coeff_determination: 0.8881
Epoch 351/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0486 - coeff_determination: 0.8832
Epoch 352/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0455 - coeff_determination: 0.8909
Epoch 353/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0487 - coeff_determination: 0.8827
Epoch 354/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0518 - coeff_determination: 0.8765
Epoch 355/500
23833/23833 [==============================] - 0s 9us/st

23833/23833 [==============================] - 0s 10us/step - loss: 0.0430 - coeff_determination: 0.8968
Epoch 417/500
23833/23833 [==============================] - 0s 10us/step - loss: 0.0418 - coeff_determination: 0.9001
Epoch 418/500
23833/23833 [==============================] - 0s 10us/step - loss: 0.0402 - coeff_determination: 0.9034
Epoch 419/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0405 - coeff_determination: 0.9033
Epoch 420/500
23833/23833 [==============================] - 0s 10us/step - loss: 0.0433 - coeff_determination: 0.8959
Epoch 421/500
23833/23833 [==============================] - 0s 10us/step - loss: 0.0415 - coeff_determination: 0.9006
Epoch 422/500
23833/23833 [==============================] - 0s 9us/step - loss: 0.0433 - coeff_determination: 0.8965
Epoch 423/500
23833/23833 [==============================] - 0s 10us/step - loss: 0.0435 - coeff_determination: 0.8962
Epoch 424/500
23833/23833 [==============================] - 0s 

23833/23833 [==============================] - 0s 8us/step - loss: 0.0376 - coeff_determination: 0.9103
Epoch 486/500
23833/23833 [==============================] - 0s 10us/step - loss: 0.0390 - coeff_determination: 0.9061
Epoch 487/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0397 - coeff_determination: 0.9046
Epoch 488/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0372 - coeff_determination: 0.9108
Epoch 489/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0378 - coeff_determination: 0.9088
Epoch 490/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0388 - coeff_determination: 0.9067
Epoch 491/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0388 - coeff_determination: 0.9072
Epoch 492/500
23833/23833 [==============================] - 0s 8us/step - loss: 0.0417 - coeff_determination: 0.9002
Epoch 493/500
23833/23833 [==============================] - 0s 8us/s

In [87]:
print(model.metrics_names)
model.evaluate(X_test, Y_test,verbose=False)

['loss', 'coeff_determination']


[0.24205226463286503, 0.30299000322748737]

In [88]:
import random
import tensorflow as tf


In [95]:
seed = round(random.uniform(1, len(x)))

X_placeholder = tf.placeholder(tf.float32, (None, X_train.shape[1])) # passing data to None column sizeis column size of train
Y_placeholder = tf.placeholder(tf.float32, (None, 1))

# define parameters
features = np.shape(X_train)[1] # switch to X_train
target_size = np.shape(X_train)[0]

learning_rate = 0.001# switch to 0.05

epochs = 1500
batch_size = 300

batch_size_placeholder = tf.placeholder(tf.int64)

# network parameters
n_hidden_1 = 50
n_hidden_2 = 25

ds_train = tf.data.Dataset.from_tensor_slices(( X_placeholder, Y_placeholder )).shuffle(buffer_size=round(len(X_train) * 0.3)).batch(batch_size_placeholder)

ds_test = tf.data.Dataset.from_tensor_slices(( X_placeholder, Y_placeholder )).batch( batch_size_placeholder )

ds_iter = tf.data.Iterator.from_structure(ds_train.output_types, ds_train.output_shapes)

next_x, next_y = ds_iter.get_next()

train_init_op = ds_iter.make_initializer(ds_train)
test_init_op = ds_iter.make_initializer(ds_test)


# initialize weights and bias  
weights = {'w1': tf.Variable(tf.random_normal([features, n_hidden_1])),
           'w2': tf.Variable(tf.random_normal([n_hidden_1, n_hidden_2])),
          'out': tf.Variable(tf.random_normal([n_hidden_2, 1]))}

biases = {'b1': tf.Variable(tf.random_normal([n_hidden_1])),
          'b2': tf.Variable(tf.random_normal([n_hidden_2])),
         'out': tf.Variable(tf.random_normal([1]))}

# Create model
def multilayer_perceptron(x, weights, biases):
    # Hidden layer 1 with ReLu activation
    layer_1 = tf.add(tf.matmul(x, weights['w1']), biases['b1'])
    layer_1 = tf.nn.relu(layer_1)
    
    # Hidden layer 2 with ReLu activation
    layer_2 = tf.add(tf.matmul(layer_1, weights['w2']), biases['b2'])
    layer_2 = tf.nn.relu(layer_2)
    
    # Output layer with ReLu activation
    out_layer = tf.add(tf.matmul(layer_2, weights['out']), biases['out'])
    return out_layer
              
# construct model
y_pred = multilayer_perceptron(next_x, weights, biases)

# define cost function(mean squred error) and optimizer(gradient descent)
cost =  tf.losses.mean_squared_error(next_y, y_pred)
optimizer = tf.train.AdamOptimizer(learning_rate=learning_rate).minimize(cost)

# initialize variables
init_op = tf.global_variables_initializer()

with tf.Session() as sess:
    sess.run(init_op)
    for epoch in range(epochs):

        sess.run(train_init_op, feed_dict={X_placeholder: X_train, Y_placeholder: Y_train, batch_size_placeholder: batch_size})
        count = 0

        while True:
            try:
                count += 1
                _, c = sess.run((optimizer, cost))

            except tf.errors.OutOfRangeError:
                break
        print('Epoch:', (epoch + 1), 'cost =', c)
    sess.run(test_init_op, feed_dict={X_placeholder: X_test, Y_placeholder: Y_test, batch_size_placeholder: len(X_test)})

    results, test_cost = sess.run((y_pred, cost))




Epoch: 1 cost = 203372.36
Epoch: 2 cost = 99795.45
Epoch: 3 cost = 52585.63
Epoch: 4 cost = 49923.39
Epoch: 5 cost = 38335.04
Epoch: 6 cost = 28421.035
Epoch: 7 cost = 24304.926
Epoch: 8 cost = 18615.873
Epoch: 9 cost = 14874.204
Epoch: 10 cost = 9836.891
Epoch: 11 cost = 8996.201
Epoch: 12 cost = 8309.597
Epoch: 13 cost = 7426.806
Epoch: 14 cost = 7535.1196
Epoch: 15 cost = 5836.95
Epoch: 16 cost = 6705.571
Epoch: 17 cost = 4009.352
Epoch: 18 cost = 4827.9795
Epoch: 19 cost = 3651.661
Epoch: 20 cost = 3341.9448
Epoch: 21 cost = 2596.146
Epoch: 22 cost = 2735.2656
Epoch: 23 cost = 3441.5986
Epoch: 24 cost = 2074.543
Epoch: 25 cost = 2005.0165
Epoch: 26 cost = 1877.6802
Epoch: 27 cost = 1300.9868
Epoch: 28 cost = 1110.1853
Epoch: 29 cost = 1086.466
Epoch: 30 cost = 1432.7711
Epoch: 31 cost = 926.0161
Epoch: 32 cost = 744.3087
Epoch: 33 cost = 876.46606
Epoch: 34 cost = 872.21954
Epoch: 35 cost = 863.8643
Epoch: 36 cost = 725.00085
Epoch: 37 cost = 466.5624
Epoch: 38 cost = 515.4012
Epoc

Epoch: 299 cost = 0.44059137
Epoch: 300 cost = 0.24003145
Epoch: 301 cost = 0.8186578
Epoch: 302 cost = 0.70546234
Epoch: 303 cost = 0.38845202
Epoch: 304 cost = 0.34998664
Epoch: 305 cost = 0.58233345
Epoch: 306 cost = 0.20697841
Epoch: 307 cost = 0.40933976
Epoch: 308 cost = 0.51548547
Epoch: 309 cost = 0.3121508
Epoch: 310 cost = 0.6330217
Epoch: 311 cost = 0.46625063
Epoch: 312 cost = 0.25466308
Epoch: 313 cost = 0.6132216
Epoch: 314 cost = 1.0498148
Epoch: 315 cost = 0.34936884
Epoch: 316 cost = 0.4147347
Epoch: 317 cost = 0.29042774
Epoch: 318 cost = 0.29417676
Epoch: 319 cost = 0.4554924
Epoch: 320 cost = 0.23983593
Epoch: 321 cost = 0.51741564
Epoch: 322 cost = 0.6259168
Epoch: 323 cost = 0.20748153
Epoch: 324 cost = 0.36273625
Epoch: 325 cost = 0.3312393
Epoch: 326 cost = 0.32351598
Epoch: 327 cost = 0.28546405
Epoch: 328 cost = 0.31104964
Epoch: 329 cost = 0.3436361
Epoch: 330 cost = 0.64095974
Epoch: 331 cost = 0.39592353
Epoch: 332 cost = 0.2915585
Epoch: 333 cost = 0.57997

Epoch: 585 cost = 0.28833878
Epoch: 586 cost = 0.47418296
Epoch: 587 cost = 0.19625907
Epoch: 588 cost = 0.23909275
Epoch: 589 cost = 0.22187962
Epoch: 590 cost = 0.1646944
Epoch: 591 cost = 0.19785409
Epoch: 592 cost = 0.21423677
Epoch: 593 cost = 0.19947313
Epoch: 594 cost = 0.26295555
Epoch: 595 cost = 0.3730266
Epoch: 596 cost = 0.2287336
Epoch: 597 cost = 0.12593096
Epoch: 598 cost = 0.18486167
Epoch: 599 cost = 0.25450888
Epoch: 600 cost = 0.29968184
Epoch: 601 cost = 0.15758277
Epoch: 602 cost = 0.23512913
Epoch: 603 cost = 0.15004468
Epoch: 604 cost = 0.26916385
Epoch: 605 cost = 0.3300918
Epoch: 606 cost = 0.4229425
Epoch: 607 cost = 0.23987482
Epoch: 608 cost = 0.21490207
Epoch: 609 cost = 0.16312538
Epoch: 610 cost = 0.19734879
Epoch: 611 cost = 0.2183158
Epoch: 612 cost = 0.23590942
Epoch: 613 cost = 0.2207718
Epoch: 614 cost = 0.15496528
Epoch: 615 cost = 0.17851844
Epoch: 616 cost = 0.1613752
Epoch: 617 cost = 0.24316418
Epoch: 618 cost = 0.17665845
Epoch: 619 cost = 0.25

Epoch: 869 cost = 0.10640504
Epoch: 870 cost = 0.17472817
Epoch: 871 cost = 0.1842423
Epoch: 872 cost = 0.08601223
Epoch: 873 cost = 0.2054173
Epoch: 874 cost = 0.14540076
Epoch: 875 cost = 0.1303482
Epoch: 876 cost = 0.11685382
Epoch: 877 cost = 0.12950866
Epoch: 878 cost = 0.16592175
Epoch: 879 cost = 0.18957865
Epoch: 880 cost = 0.14688188
Epoch: 881 cost = 0.12095414
Epoch: 882 cost = 0.15117325
Epoch: 883 cost = 0.15369517
Epoch: 884 cost = 0.1420738
Epoch: 885 cost = 0.16447367
Epoch: 886 cost = 0.16591345
Epoch: 887 cost = 0.13054024
Epoch: 888 cost = 0.21595578
Epoch: 889 cost = 0.1848813
Epoch: 890 cost = 0.10242734
Epoch: 891 cost = 0.261161
Epoch: 892 cost = 0.14347053
Epoch: 893 cost = 0.14406934
Epoch: 894 cost = 0.08823003
Epoch: 895 cost = 0.1084421
Epoch: 896 cost = 0.15932095
Epoch: 897 cost = 0.13824643
Epoch: 898 cost = 0.19198403
Epoch: 899 cost = 0.1400577
Epoch: 900 cost = 0.12774393
Epoch: 901 cost = 0.1506194
Epoch: 902 cost = 0.15057617
Epoch: 903 cost = 0.1633

Epoch: 1147 cost = 0.106047764
Epoch: 1148 cost = 0.25511402
Epoch: 1149 cost = 0.14180174
Epoch: 1150 cost = 0.13542438
Epoch: 1151 cost = 0.082744144
Epoch: 1152 cost = 0.07653659
Epoch: 1153 cost = 0.0895909
Epoch: 1154 cost = 0.10037676
Epoch: 1155 cost = 0.11595552
Epoch: 1156 cost = 0.09797552
Epoch: 1157 cost = 0.18206438
Epoch: 1158 cost = 0.17505404
Epoch: 1159 cost = 0.12897941
Epoch: 1160 cost = 0.1294497
Epoch: 1161 cost = 0.07849912
Epoch: 1162 cost = 0.06806501
Epoch: 1163 cost = 0.08953619
Epoch: 1164 cost = 0.09293495
Epoch: 1165 cost = 0.11369951
Epoch: 1166 cost = 0.12270079
Epoch: 1167 cost = 0.07986301
Epoch: 1168 cost = 0.12133741
Epoch: 1169 cost = 0.093833625
Epoch: 1170 cost = 0.11763325
Epoch: 1171 cost = 0.1638316
Epoch: 1172 cost = 0.10756123
Epoch: 1173 cost = 0.115904726
Epoch: 1174 cost = 0.07714416
Epoch: 1175 cost = 0.12291986
Epoch: 1176 cost = 0.13687268
Epoch: 1177 cost = 0.1832274
Epoch: 1178 cost = 0.116355754
Epoch: 1179 cost = 0.13216802
Epoch: 11

Epoch: 1420 cost = 0.13148846
Epoch: 1421 cost = 0.18057455
Epoch: 1422 cost = 0.0800193
Epoch: 1423 cost = 0.13052714
Epoch: 1424 cost = 0.102360055
Epoch: 1425 cost = 0.14038898
Epoch: 1426 cost = 0.13178824
Epoch: 1427 cost = 0.12230808
Epoch: 1428 cost = 0.15447593
Epoch: 1429 cost = 0.110703595
Epoch: 1430 cost = 0.0954577
Epoch: 1431 cost = 0.13333166
Epoch: 1432 cost = 0.0885504
Epoch: 1433 cost = 0.06769792
Epoch: 1434 cost = 0.12794285
Epoch: 1435 cost = 0.117945455
Epoch: 1436 cost = 0.108549625
Epoch: 1437 cost = 0.11879992
Epoch: 1438 cost = 0.16110758
Epoch: 1439 cost = 0.09581028
Epoch: 1440 cost = 0.049800515
Epoch: 1441 cost = 0.066159405
Epoch: 1442 cost = 0.08190737
Epoch: 1443 cost = 0.08602562
Epoch: 1444 cost = 0.16919962
Epoch: 1445 cost = 0.11805666
Epoch: 1446 cost = 0.079511225
Epoch: 1447 cost = 0.104383945
Epoch: 1448 cost = 0.10140211
Epoch: 1449 cost = 0.08975141
Epoch: 1450 cost = 0.10713098
Epoch: 1451 cost = 0.13153379
Epoch: 1452 cost = 0.11927218
Epoch

In [96]:
#    print(results)
    print(test_cost)
    print('R^2:', r_square(np.reshape(results, (len(results),)),Y_test))

0.4086066
5959 [1599.30598938] 3152.5502960970407 [2467.37678109]
R^2: [0.32882561]
